harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -425,43 +425,49 @@ def _compute_validation_metrics(self) -> workload.Response:
         metrics = {}  # type: Optional[Dict[str, Any]]
 
         if self._evaluate_batch_defined():
-            keys = None
-            batch_metrics = []
+            keys = None  # type: Optional[KeysView[str]]
+            reducers = {}
+            last_state = {}  # type: Dict[str, Any]
 
             self.validation_loader = cast(torch.utils.data.DataLoader, self.validation_loader)
             check.gt(len(self.validation_loader), 0)
             for batch in self.validation_loader:
                 batch = self.context.to_device(batch)
                 num_inputs += pytorch._data_length(batch)
 
-                vld_metrics = self.trial.evaluate_batch(batch=batch)
+                vld_metrics = self.trial.evaluate_batch(batch)
                 # Verify validation metric names are the same across batches.
                 if keys is None:
                     keys = vld_metrics.keys()
+                    reducers = self._prepare_metrics_reducers(keys)
                 else:
                     check.eq(
                         keys,
                         vld_metrics.keys(),
                         "Validation metric names must match across all batches of data.",
                     )
-                check.is_instance(
-                    vld_metrics,
-                    dict,
-                    "validation_metrics() must return a "
-                    "dictionary of string names to Tensor "
-                    "metrics",
-                )
                 # TODO: For performance perform -> cpu() only at the end of validation.
-                batch_metrics.append(self._convert_metrics_to_numpy(vld_metrics))
+                # XXX: figure this out, as it will be part of the reducers API.
```

Comment:
I use `XXX` to remind myself that something must be fixed before landing a PR.  In this case I'm referring to the above line:

```
# TODO: For performance perform -> cpu() only at the end of validation.
```

where we need to actually decide now if we want to have the reducers operate on the device or on the CPU, because it would affect user code.  I'm leaning towards having it happen on the device and calling `.numpy()` after the reducer is done